cli/flox-rust-sdk/src/providers/flox_cpp_utils.rs

```diff
@@ -0,0 +1,281 @@
+/// time-consuming unless cached.
+///
+/// The trait is implemented by the [`Pkgdb`] struct which is the canonical implementation
+/// using the `pkgdb lock-flake-installable` command.
+///
+/// The trait is also implemented by the [`LockFlakeInstallableMock`] struct which is used for testing.
+#[enum_dispatch]
+pub trait LockFlakeInstallableTrait {
+    fn lock_flake_installable(
+        &self,
+        system: impl AsRef<str>,
+        installable: impl AsRef<str>,
+    ) -> Result<LockedInstallable, FlakeInstallableError>;
+}
+
+#[derive(Debug)]
+#[enum_dispatch(LockFlakeInstallableTrait)]
+pub enum LockFlakeInstallable {
+    Pkgdb(Pkgdb),
+    Mock(LockFlakeInstallableMock),
+}
+
+impl Default for LockFlakeInstallable {
+    fn default() -> Self {
+        LockFlakeInstallable::Pkgdb(Pkgdb::default())
+    }
+}
+/// A wrapper for (eventually) various `pkgdb` commands
+/// Currently only implements [LockFlakeInstallableTrait] through
+/// `pkgdb lock-flake-installable`.
+#[derive(Debug)]
+pub struct Pkgdb {
+    pkgdb_path: PathBuf,
+}
+
+impl Pkgdb {
+    fn new() -> Self {
+        Pkgdb {
+            pkgdb_path: PathBuf::from(&*PKGDB_BIN),
+        }
+    }
+}
+
+impl Default for Pkgdb {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl LockFlakeInstallableTrait for Pkgdb {
+    fn lock_flake_installable(
+        &self,
+        system: impl AsRef<str>,
+        installable: impl AsRef<str>,
+    ) -> Result<LockedInstallable, FlakeInstallableError> {
+        let installable = installable.as_ref();
+        let mut pkgdb_cmd = Command::new(&self.pkgdb_path);
+
+        pkgdb_cmd
+            .arg("lock-flake-installable")
+            .args(["--system", system.as_ref()])
+            .arg(installable);
+
+        debug!("locking installable: {pkgdb_cmd:?}");
+
+        let lock = call_pkgdb(pkgdb_cmd).map_err(|err| match err {
+            CallPkgDbError::PkgDbError(PkgDbError {
+                exit_code: error_codes::NIX_LOCK_FLAKE,
+                context_message:
+                    Some(ContextMsgError {
+                        caught: Some(nix_error),
+                        ..
+                    }),
+                ..
+            }) => FlakeInstallableError::LockInstallable(nix_error.message),
+            _ => FlakeInstallableError::Pkgdb(err),
+        })?;
+
+        let lock = serde_json::from_value(lock)
+            .map_err(FlakeInstallableError::DeserializeLockedInstallable)?;
+
+        Ok(lock)
+    }
+}
+
+/// Mock implementation of [`LockFlakeInstallableTrait`] for testing.
+#[derive(Debug, Default)]
+pub struct LockFlakeInstallableMock {
+    lock_flake_installable: Arc<Mutex<VecDeque<Result<LockedInstallable, FlakeInstallableError>>>>,
+}
+
+impl LockFlakeInstallableMock {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[allow(unused)]
+    fn push(&self, result: Result<LockedInstallable, FlakeInstallableError>) {
```

Comment:
```suggestion
    fn push_locked_installable(&self, result: Result<LockedInstallable, FlakeInstallableError>) {
```
It should be very obvious at the call site what's being pushed or what you're expected to be pushing.

Reply:
I mean, isn't it obvious from the type signature?
But since it does take a result maybe `push_lock_result` would be clearer even i we want to make the name carry more information?